test/Performance/BuildPerformanceTest.cs

```diff
@@ -0,0 +1,285 @@
+
+        [Benchmark]
+        public void IncrementalSkipAllNoDependenciesInGraph_TwoTargetGraph() =>
+            IncrementalSkipAllNoDependenciesInGraph(CreateTestInstances(TwoTargetGraph));
+
+        [Benchmark]
+        public void IncrementalSkipAllNoDependenciesInGraph_TwoTargetGraphLarge() =>
+            IncrementalSkipAllNoDependenciesInGraph(CreateTestInstances(TwoTargetGraphLarge));
+
+        public void IncrementalSkipAllNoDependenciesInGraph(TestInstance[] instances)
+        {
+            var instance = instances[0];
+            new BuildCommand(instance.TestRoot, buildProfile: false)
+                   .Execute().Should().Pass();
+
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                {
+                    foreach (var i in instances)
+                    {
+                        new BuildCommand(i.TestRoot,
+                            framework: DefaultFramework,
+                            noDependencies: true,
+                            buildProfile: false)
+                            .Execute().Should().Pass();
+                    }
+                }
+            }
+        }
+        [Benchmark]
+        public void BuildAllNoDependenciesInGraph_SingleTargetGraph() =>
+          BuildAllNoDependenciesInGraph(CreateTestInstances(SingleTargetGraph));
+
+        [Benchmark]
+        public void BuildAllNoDependenciesInGraph_TwoTargetGraph() =>
+            BuildAllNoDependenciesInGraph(CreateTestInstances(TwoTargetGraph));
+
+        [Benchmark]
+        public void BuildAllNoDependenciesInGraph_TwoTargetGraphLarge() =>
+            BuildAllNoDependenciesInGraph(CreateTestInstances(TwoTargetGraphLarge));
+
+        public void BuildAllNoDependenciesInGraph(TestInstance[] instances)
+        {
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                {
+                    foreach (var i in instances.Reverse())
+                    {
+                        new BuildCommand(i.TestRoot,
+                            framework: DefaultFramework,
+                            noDependencies: true,
+                            buildProfile: false)
+                            .Execute().Should().Pass();
+                    }
+                }
+                foreach (var instance in instances)
+                {
+                    TouchSource(instance.TestRoot);
+                }
+            }
+        }
+
+        protected void TouchSource(string project)
+        {
+            var sourceFile = Directory.GetFiles(project, "*.cs").FirstOrDefault();
+            if (sourceFile == null)
+            {
+                throw new InvalidOperationException($"'.cs' files not found in {project}");
+            }
+            File.SetLastWriteTime(sourceFile, DateTime.Now);
```

Comment:
I've not tried updating the last write time.  However, if this is how both the compiler and the CLI decide whether to compile something, then I think this is fine - you don't need to actually mutate the source.